From 0c845ae43c5dd86ad868db5886e6afbe74e50a73 Mon Sep 17 00:00:00 2001
From: dann frazier <dann.frazier@chainguard.dev>
Date: Thu, 14 Aug 2025 12:36:51 -0600
Subject: [PATCH] Add a distutils option --install-layout=chainguard

Based on Debian's distutils-install-layout.diff

This option:
  - installs into $prefix/dist-packages instead of $prefix/site-packages.
  - doesn't encode the python version into the egg name.

 We install modules into dist-packages to avoid entangling native OS
 packages with pip-installed packages.
 .
 Customize site.py to import from Chainguard's dist-packages layout.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>
---
 Lib/pydoc.py          |  1 +
 Lib/site.py           | 16 +++++++++++++++-
 Lib/test/test_site.py | 10 +++++-----
 3 files changed, 21 insertions(+), 6 deletions(-)

diff --git a/Lib/pydoc.py b/Lib/pydoc.py
index d508fb70ea4..ff2603c7e35 100644
--- a/Lib/pydoc.py
+++ b/Lib/pydoc.py
@@ -577,6 +577,7 @@ def getdocloc(self, object, basedir=sysconfig.get_path('stdlib')):
                                  'marshal', 'posix', 'signal', 'sys',
                                  '_thread', 'zipimport') or
              (file.startswith(basedir) and
+              not file.startswith(os.path.join(basedir, 'dist-packages')) and
               not file.startswith(os.path.join(basedir, 'site-packages')))) and
             object.__name__ not in ('xml.etree', 'test.test_pydoc.pydoc_mod')):
             if docloc.startswith(("http://", "https://")):
diff --git a/Lib/site.py b/Lib/site.py
index f9327197159..02b04b122c2 100644
--- a/Lib/site.py
+++ b/Lib/site.py
@@ -7,12 +7,17 @@
 This will append site-specific paths to the module search path.  On
 Unix (including Mac OSX), it starts with sys.prefix and
 sys.exec_prefix (if different) and appends
-lib/python<version>/site-packages.
+lib/python<version>/dist-packages.
 On other platforms (such as Windows), it tries each of the
 prefixes directly, as well as with lib/site-packages appended.  The
 resulting directories, if they exist, are appended to sys.path, and
 also inspected for path configuration files.
 
+For Chainguard, this sys.path is augmented with directories
+for packages distributed within the distribution. Chainguard addons
+install into /usr/lib/python<version>/dist-packages.
+/usr/lib/python<version>/site-packages is not used.
+
 If a file named "pyvenv.cfg" exists one directory above sys.executable,
 sys.prefix and sys.exec_prefix are set to that directory and
 it is also checked for site-packages (sys.base_prefix and
@@ -394,6 +399,8 @@ def getsitepackages(prefixes=None):
     if prefixes is None:
         prefixes = PREFIXES
 
+    is_virtual_environment = sys.base_prefix != sys.prefix or hasattr(sys, "real_prefix")
+
     for prefix in prefixes:
         if not prefix or prefix in seen:
             continue
@@ -410,6 +417,13 @@ def getsitepackages(prefixes=None):
             if sys.platlibdir != "lib":
                 libdirs.append("lib")
 
+            if is_virtual_environment:
+                sitepackages.append(os.path.join(prefix, "lib",
+                                                 "python%d.%d" % sys.version_info[:2],
+                                                 "site-packages"))
+            sitepackages.append(os.path.join(prefix, "lib",
+                                             "python%d.%d" % sys.version_info[:2],
+                                             "dist-packages"))
             for libdir in libdirs:
                 path = os.path.join(prefix, libdir,
                                     f"{implementation}{ver[0]}.{ver[1]}{abi_thread}",
diff --git a/Lib/test/test_site.py b/Lib/test/test_site.py
index d0e32942635..312bf39ac43 100644
--- a/Lib/test/test_site.py
+++ b/Lib/test/test_site.py
@@ -326,16 +326,16 @@ def test_getsitepackages(self):
         if os.sep == '/':
             # OS X, Linux, FreeBSD, etc
             if sys.platlibdir != "lib":
-                self.assertEqual(len(dirs), 2)
+                self.assertEqual(len(dirs), 3)
                 wanted = os.path.join('xoxo', sys.platlibdir,
-                                      f'python{sysconfig._get_python_version_abi()}',
-                                      'site-packages')
+                                  f'python{sysconfig._get_python_version_abi()}',
+                                  'site-packages')
                 self.assertEqual(dirs[0], wanted)
             else:
-                self.assertEqual(len(dirs), 1)
+                self.assertEqual(len(dirs), 3)
             wanted = os.path.join('xoxo', 'lib',
                                   f'python{sysconfig._get_python_version_abi()}',
-                                  'site-packages')
+                                  'dist-packages')
             self.assertEqual(dirs[-1], wanted)
         else:
             # other platforms
-- 
2.48.1

